Add support for custom messages to restricted imports

majapw opened the pull request:
**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[X] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

**What rule do you want to change?**
`no-restricted-imports` and `no-restricted-modules` (am interested in feedback first before making the changes to `no-restricted-modules` as well)

**Does this change cause the rule to produce more or fewer warnings?**
Same amount of errors but adds more nuance to the error messaging

**How will the change be implemented? (New option, new default behavior, etc.)?**
Modification to allowed options. Anywhere you can pass in a path string, you can now also pass in an object with a `name` and `message` key where the `name` corresponds to a path and the `message` is the custom message you want appended to the default error message.

**Please provide some example code that this change will affect:**
All existing configurations will continue to work. However, the following will also be accepted:
```json
"no-restricted-imports": ["error", [{
  "name": "import-foo",
  "message": "Please use import-bar instead."
}]]
```

or like this:

```json
"no-restricted-imports": ["error", {
  "paths": [{
    "name": "import-foo",
    "message": "Please use import-bar instead."
  }]
}]
```

**What does the rule currently do for this code?**
errors on any restricted imports that are passed into the options.

**What will the rule do after it's changed?**
continue to error on any restricted imports that are passed into the options, but with more clarity on what the developer should do instead.

**What changes did you make? (Give an overview)**
Following the precedent set by no-restricted-globals, anywhere you can pass a path string, you can also pass an object with a "name" and "message" key. This custom message will be appended to any error message triggered by an import of the "name" value. Custom messages are not supported with patterns.

**Is there anything you'd like reviewers to focus on?**
See if this pattern (and naming scheme) is acceptable so as to do the same for `no-restricted-modules`

FYI @ljharb @yzimet

not-an-aardvark commented:
Thanks for the PR. Have you seen https://github.com/eslint/eslint/issues/8400? It looks like a similar proposal was accepted to add both `paths` and `patterns` options to `no-restricted-imports`, to allow for custom messages to be used.

I think it would be fine to just add the `paths` option in this PR, but I want to make sure this PR is compatible with the plan in https://github.com/eslint/eslint/issues/8400 so that we would still be able to add `patterns` later.

ljharb commented:
The plan for "paths" is identical in this PR to #8400; I definitely think this could be merged right now, and in no way would it block #8400.

not-an-aardvark commented on lib/rules/no-restricted-imports.js:
```diff
@@ -47,35 +71,77 @@ module.exports = {
+        function reportPath(node) {
+            const importName = node.source.value.trim();
+            const customMessage = restrictedPathMessages[importName];
+            const message = customMessage
```
This will cause the default message to get used if an empty string gets used as the custom message, which I'm assuming is unintended. (It's unlikely that users will provide an empty string very often, but I think it would be better to handle this case consistently with the case where non-empty strings are provided.)

ljharb replied:
Could that be resolved by modifying the JSON schema to forbid an empty message?

ljharb replied:
(specifically, by adding `minLength: 1`)

not-an-aardvark replied:
That would also work. I don't have any strong preferences for how empty strings are handled; I just want to make sure that the behavior is consistent with the rest of the rule (or that if the behavior is inconsistent, it's a design choice rather than a bug).

not-an-aardvark commented on lib/rules/no-restricted-imports.js:
```diff
@@ -47,35 +71,77 @@ module.exports = {
+         * @private
+         */
+        function isRestrictedPath(name) {
+            return restrictedPathMessages.hasOwnProperty(name);
```
I think this will cause an error to get thrown if someone uses the string `"hasOwnProperty"` as a path. It would be better to use something like `Object.prototype.hasOwnProperty.call(restrictedPathMessages, name)`.

majapw commented:
@not-an-aardvark I think I've addressed your concerns (`hasOwnProperty` on the `restrictedPathMessages` object and add a `minLength` to the schema for the custom message). I've also duplicated the effort for `no-restricted-modules`. Please take another look!

majapw commented on docs/rules/no-restricted-modules.md:
````diff
@@ -26,6 +26,28 @@ For example, to restrict the use of all Node.js core modules (via https://github
+You may also specify a custom message for any paths you want to restrict as follows:
+
+```json
+"no-restricted-imports": ["error", [{
````
Oh whoops! My copy-pasta bad

majapw commented:
@not-an-aardvark addressed!

platinumazure commented on docs/rules/no-restricted-imports.md:
````diff
@@ -35,6 +35,28 @@ When using the object form, you can also specify an array of gitignore-style pat
+}]
+```
+
+The custom message will be appends to the default error message. Please note that you may not specify custom error messages for restricted patterns as a particular import may match more than one pattern.
````
I think this should say "The custom message will be appended to the default error message."

platinumazure commented on docs/rules/no-restricted-modules.md:
````diff
@@ -26,6 +26,28 @@ For example, to restrict the use of all Node.js core modules (via https://github
+}]
+```
+
+The custom message will be appends to the default error message. Please note that you may not specify custom error messages for restricted patterns as a particular module may match more than one pattern.
````
I think this should say "The custom message will be appended to the default error message."

platinumazure commented on lib/rules/no-restricted-imports.js:
```diff
@@ -47,35 +74,77 @@ module.exports = {
+        }
+
+        /**
+         * Check if the given name is a restricted global name.
```
Could this comment be edited? (This seems to have been copied from no-restricted-globals)

platinumazure commented on lib/rules/no-restricted-modules.js:
```diff
@@ -76,6 +116,39 @@ module.exports = {
+        }
+
+        /**
+         * Check if the given name is a restricted global name.
```
Could this comment be edited? (This seems to have been copied from no-restricted-globals)

kaicataldo commented:
Let's make sure to update the issue after this lands so that it's clear that only custom messages for `patterns` are needed.

majapw commented:
@platinumazure I believe I've addressed your comments as well.

kaicataldo commented:
Thanks for contributing to ESLint!
